Initial TIS-CI configuration

jakub-zwolakowski opened the pull request:
Verifying all test vectors with TIS-CI for four architectures: x86 and ppc, 32-bit and 64-bit.

Feel free to merge it to a test branch (like **tis-ci**) before deciding to put it on master.

About the files that are included:

*   `tis.config` :  
    The main file describing the execution paths to analyze (i.e. your tests) and the analysis options for all the cases.  
    (For more information see: [https://docs.ci.trust-in-soft.com/reference/configuration-file](https://docs.ci.trust-in-soft.com/reference/configuration-file))
*   `c/gen_tis.config.py` :  
    The test script `test.py`, modified to 1) extract all the test vector data and put it into separate files in the directory `tis-ci/test_vectors`, 2) generate the `tis.config` file that mirrors all the tests that are usually ran by `test.py`.  
    (This script is a prototype solution which contains a lot of redundancy with `test.py`. The project's maintainers should have a better idea how to handle this, e.g. leave it as a separate script, incorporate into `test.py`, etc.)
*   `tis-ci/test.c` :  
    A test driver which wraps the main function in order to additionally check whether the test results are correct and whether the main function's return value is zero.  
    (This is optional. Usually TIS-CI only checks if there are no UB on the execution path as checking test results is redundant: the normal test suite is supposed to already handle this. However, it happened to be pretty easy and quick to add the test output checks for BLAKE3, also it proved helpful as it has let us discover some wrong test results on big-endian architectures: [https://github.com/BLAKE3-team/BLAKE3/pull/118](https://github.com/BLAKE3-team/BLAKE3/pull/118))
*   `tis-ci/test_vectors` :  
    The data extracted from test\_vectors/test\_vectors.json to separate files which can be thus be used as input.
*   Also `c/main.c` is slightly modified in order to facilitate the test results checking done by tis-ci/test.c.
*   And the TIS-CI badge was added to `README.md`. It shows the verification status for the last commit on the **master** branch, if you want something else (e.g. last commit on any branch) you'd have to modify it.  
    (For more information see: [https://docs.ci.trust-in-soft.com/reference/status-badge](https://docs.ci.trust-in-soft.com/reference/status-badge))

After merging this PR, you only need to add the repo to TIS-CI from your side. A tutorial is available ([https://docs.ci.trust-in-soft.com/tutorial/set-up-the-continuous-verification#trigger-the-first-analysis](https://docs.ci.trust-in-soft.com/tutorial/set-up-the-continuous-verification#trigger-the-first-analysis)), but it boils down to these steps:

1.  Go to [https://ci.trust-in-soft.com/](https://ci.trust-in-soft.com/) and click on **Sign-up with GitHub**,
2.  During the sign-up process don't forget to grant TIS-CI access to the **BLAKE3-team** organization,
3.  If you already signed up and TIS-CI doesn’t have access to **BLAKE3-team**, you can follow these steps to grant access:  
    [https://docs.ci.trust-in-soft.com/reference/github-organizations#after-sign-up](https://docs.ci.trust-in-soft.com/reference/github-organizations#after-sign-up)
4.  Once in **TIS-CI**, click on **Add Project** (in the top navigation menu),
5.  Choose **BLAKE3-team** from the organization list, then **BLAKE3** from the project list,
6.  Choose the branch where this PR was merged (the one with the `tis.config` file),
7.  Click on **Start verifying**.

The final effect should look similar to this: [https://ci.trust-in-soft.com/projects/jakub-zwolakowski/BLAKE3/6?page=1](https://ci.trust-in-soft.com/projects/jakub-zwolakowski/BLAKE3/6?page=1)

From this point forward all pushes to the repo will automatically trigger a new TIS-CI verification on all the branches which contain a `tis.config` file at the repository root. And you will stay up-to-date on the results thanks to this TIS-CI badge in your README.md.

So after this you should be all good!

jakub-zwolakowski committed:
Initial TIS-CI configuration, four architectures: x86 and ppc, 32-bit and 64-bit.

oconnor663 commented on README.md:
```diff
@@ -63,6 +63,8 @@ This repository is the official implementation of BLAKE3. It includes:
 
 * [![Actions Status](https://github.com/BLAKE3-team/BLAKE3/workflows/tests/badge.svg)](https://github.com/BLAKE3-team/BLAKE3/actions)
 
+* [![TrustInSoft CI](https://ci.trust-in-soft.com/projects/BLAKE3-team/BLAKE3.svg?branch=master)](https://ci.trust-in-soft.com/projects/BLAKE3-team/BLAKE3)
```
I have a slight preference for merging this line with the one above, to keep the testing badges together.

jakub-zwolakowski replied:
Done!

oconnor663 commented on c/main.c:
```diff
@@ -70,6 +70,10 @@ enum cpu_feature {
+#ifdef __TRUSTINSOFT_ANALYZER__
+extern FILE *fp_output;
+#endif
+
```
I've been thinking about simplifying this `main.c` file for a while. A lot of different tests flow through here, and as a result it does some weird stuff like buffering its entire input in memory and writing to the private global `g_cpu_features` flags. Originally I wrote it to be a usable by a human on the command line, but at some point I gave up on that. Eventually I'd like to take it apart, but with this change I'd be worried about breaking our TIS tests and not knowing how to fix them.

Could it make sense to just copy the important parts into `tis-ci/test.c`? If the TIS tests need private access to e.g. `g_cpu_features`, I wonder if we could add preprocessor flags or maybe new API functions to support them.

jakub-zwolakowski replied:
Hmm. Basically: do as you please, as you surely know much better than me how this code should be organized :)

To give you some context: my aim here was to deploy TIS-CI on the project in a simple and efficient way, piggy-backing on existing test-suite as much as possible. So frankly I did not even look too much into what was exactly going on in `main.c` and why. I've used it pretty much as-is, because it is the entry point for these test vector tests from `c/test.py`. The only problem that I've encountered was that TIS-CI couldn't compare what was printed on stdout with the expected output read from a file. Hence the only change that I needed to implement was adding the printing to an external file in addition to printing to stdout.

So, answering your question: yes, as long as there exists a function that can be called from `main_wrapper` that outputs the same thing as current `main` in `c/main.c` (to an external file and not only stdout) and which returns the same values, then TIS-CI will be fine. You can move or copy or trim this code as you see fit, e.g. putting it directly in `tis-ci/test.c` or in a new file `tis-ci/main.c`, whichever seems easier to maintain (some minor changes to `tis.config` might be necessary you move the files around, but I can help you with that).

jakub-zwolakowski committed:
Keep the testing badges together.

oconnor663 commented:
(Apologies for sitting on this. I'll make some time for it this weekend.)

jakub-zwolakowski commented:
Hi @oconnor663 ! Did you find some time to look at the setup? I'm here to help if needed :)

oconnor663 commented:
Forgive me, I didn't. I've put it back on the calendar for this coming weekend.

oconnor663 commented:
As I go through the flow you mention above, I notice that it requires me to give TIS commit privileges on this repo. Is there a specific reason that's necessary?

oconnor663 commented:
Also if I'm making changes to the test code here, is there a way for me to try it out locally? Or is the best flow to just let the TIS website test the branch?

jakub-zwolakowski commented:
> As I go through the flow you mention above, I notice that it requires me to give TIS commit privileges on this repo. Is there a specific reason that's necessary?

Yes, there is a reason. TIS CI will never write to your repo, yet unfortunately it has to ask for write privileges even if it only reads, it's a known GitHub API limitation.

See https://docs.ci.trust-in-soft.com/tutorial/set-up-the-continuous-verification#sign-up-to-trustinsoft-ci

jakub-zwolakowski commented:
> Also if I'm making changes to the test code here, is there a way for me to try it out locally? Or is the best flow to just let the TIS website test the branch?

Sorry, currently TIS-CI is a 100% online service, so you cannot try things locally. You need to trigger a verification, i.e. push a commit that contains the `tis.config` file at the repository's root to any branch. You can always remove the spurious build later on from TIS CI if this bothers you (and / or rewrite your own repo's history on the GitHub side by resetting the branch to a previous commit and force pushing).

jakub-zwolakowski committed:
Improve the robustness of the TIS CI config.

jakub-zwolakowski commented:
Hi @oconnor663 !

I've made the initial TIS CI configuration more robust:

- The script `trustinsoft/regenerate.py` regenerates all the TrustInSoft CI configuration files as well as the test inputs and expected outputs which TIS CI uses (from the test vectors JSON `test_vectors/test_vectors.json`). The script still contains bits of logic from `c/test.py`, I've marked these parts with comments. Maybe a separate file with these common parts could be created one day, but for now I didn't want to be too intrusive in the existing code, so I've decided that copying these few lines was more appropriate.
- A large part of `c/main.c` (a bit simplified, e.g. I removed the usage of `g_cpu_features`) was incorporated directly into `trustinsoft/test.c`, so we are not relying on that file anymore. If I understood well, this was one of your major concerns.

I hope this addresses some of your concerns :)
